### PR TITLE
Bug fix: Emit code for C++ inline function called from another inline function

### PR DESCRIPTION
**Pull request by martinboehme**
Adds a test that fails without the fix.

The fix consists simply of removing a return statement; as a result, we now always emit code for functions referenced from the function we are currently emitting.

Previously, this was not done for externally visible functions. This goes all the way back to when `GenClangDecl.cpp` was originally introduced in https://github.com/apple/swift/commit/ee22004b84dd72fd9d99dbb11837f5dfd377d08a.

I speculate that the reason for doing this was that in C and Objective-C, we can assume that an externally visible function will be provided by some `.o` file we're linking against. This is also true for `inline` functions in C99, where exactly one translation unit needs to declare the function `extern inline` to cause it to get emitted to the `.o` file.

C++ inline functions follow a different model, however, so we need to make sure we emit code for them.

### CODE REVIEW

File: test/Interop/Cxx/function/Inputs/module.modulemap [hunk 1]
```diff
@@ -0,0 +1,4 @@
+module EmitCalledFunction {
+  header "EmitCalledFunction.h"
```
Comment (MForster) [inline test/Interop/Cxx/function/Inputs/module.modulemap:2]: (nit) Should we keep to a consistent naming convention for files (lowercase with dashes)?

**Comment by gribozavr** [comment]
@swift-ci Please smoke test

File: test/Interop/Cxx/function/emit-called-function-ir.swift [hunk 2]
```diff
@@ -0,0 +1,14 @@
+// This is a reproducer for a bug where Swift was not emitting code for a
```
Comment (gribozavr) [inline test/Interop/Cxx/function/emit-called-function-ir.swift:1]: Could you rephrase this as a test for emitting inline functions that are used in general? The "regression test" wording is useful for obscure combinations, but this seems like a pretty basic case that we should be testing regardless.

Comment (gribozavr) [inline test/Interop/Cxx/function/emit-called-function-ir.swift:1]: s/-ir/-irgen/ in the file name.

File: test/Interop/Cxx/function/Inputs/module.modulemap (hunk 1 above)
Comment (martinboehme) [reply test/Interop/Cxx/function/Inputs/module.modulemap]: Good point -- done.

File: test/Interop/Cxx/function/Inputs/EmitCalledFunction.h [hunk 3]
```diff
@@ -0,0 +1,7 @@
+inline int bar() {
+  return 42;
+}
+
+inline int foo() {
+  return bar();
+}
```
Comment (gribozavr) [inline test/Interop/Cxx/function/Inputs/EmitCalledFunction.h]: Please use more descriptive names than foo and bar.

File: test/Interop/Cxx/function/emit-called-function-ir.swift (hunk 2 above)
Comment (martinboehme) [reply test/Interop/Cxx/function/emit-called-function-ir.swift]: Good point, done.

Comment (martinboehme) [reply test/Interop/Cxx/function/emit-called-function-ir.swift]: Done.

File: test/Interop/Cxx/function/Inputs/EmitCalledFunction.h (hunk 3 above)
Comment (martinboehme) [reply test/Interop/Cxx/function/Inputs/EmitCalledFunction.h]: Done.

While I was doing this, it occurred to me that I should also add a function that isn't called from Swift at all (whether directly or indirectly) and test that it isn't emitted.

File: test/Interop/Cxx/function/emit-called-function-irgen.swift [hunk 4]
```diff
@@ -0,0 +1,20 @@
+// Test that Swift emits code for a C++ inline function that is called from
```
Comment (gribozavr) [inline test/Interop/Cxx/function/emit-called-function-irgen.swift:1]: Please add "inline" to the names of the tests.

**Comment by martinboehme** [comment]
@rjmccall I'd like your feedback on whether this fix looks right.

File: test/Interop/Cxx/function/emit-called-function-irgen.swift [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+//
+// We should not, however, emit code for functions that aren't called even
+// indirectly from Swift.
+
+// RUN: %empty-directory(%t)
+// RUN: %target-swift-frontend %s -I %S/Inputs -enable-cxx-interop -emit-ir -o - | %FileCheck %s
+
+import EmitCalledFunction
+
+// Unfortunately, we have to repeat the CHECK-NOT to express that notCalled()
```
Comment (gribozavr) [inline test/Interop/Cxx/function/emit-called-function-irgen.swift:12]: Please use `--implicit-check-not`.

File: test/Interop/Cxx/function/emit-called-function-irgen.swift (hunk 4 above)
Comment (gribozavr) [inline test/Interop/Cxx/function/emit-called-function-irgen.swift:1]: "Test that we emit LLVM IR for C++ inline free functions that are called directly, and also for inline functions that are called transitively. Test that we don't emit LLVM IR for inline functions that are not called."

File: test/Interop/Cxx/function/emit-called-function-irgen.swift [hunk 6]
```diff
@@ -0,0 +1,20 @@
+// Test that Swift emits code for a C++ inline function that is called from
+// another C++ inline function, but not directly from Swift code.
+//
+// We should not, however, emit code for functions that aren't called even
+// indirectly from Swift.
+
+// RUN: %empty-directory(%t)
+// RUN: %target-swift-frontend %s -I %S/Inputs -enable-cxx-interop -emit-ir -o - | %FileCheck %s
```
Comment (gribozavr) [inline test/Interop/Cxx/function/emit-called-function-irgen.swift:8]: WDYT about running this test in both C and C++ modes?

File: test/Interop/Cxx/function/Inputs/emit-called-function.h [hunk 7]
```diff
@@ -0,0 +1,11 @@
+inline int notCalled() {
+  return 42;
+}
+
+inline int calledIndirectly() {
```
Comment (gribozavr) [inline test/Interop/Cxx/function/Inputs/emit-called-function.h:5]: s/indirectly/transitively/

"Indirectly" means "through a pointer".

File: test/Interop/Cxx/function/emit-called-function-irgen.swift (hunk 4 above)
Comment (martinboehme) [reply test/Interop/Cxx/function/emit-called-function-irgen.swift]: Done.

File: test/Interop/Cxx/function/emit-called-function-irgen.swift (hunk 5 above)
Comment (martinboehme) [reply test/Interop/Cxx/function/emit-called-function-irgen.swift]: Done.

File: test/Interop/Cxx/function/emit-called-function-irgen.swift (hunk 4 above)
Comment (martinboehme) [reply test/Interop/Cxx/function/emit-called-function-irgen.swift]: Thanks, this is much clearer. Done, with some additional wordsmithing that I hope you're OK with.

File: test/Interop/Cxx/function/emit-called-function-irgen.swift (hunk 6 above)
Comment (martinboehme) [reply test/Interop/Cxx/function/emit-called-function-irgen.swift]: Good point. Done.

I'm using `static inline` in C, as `extern inline` follows a [different model](https://gustedt.wordpress.com/2010/11/29/myth-and-reality-about-inline-in-c99/), where there's always one translation unit that explicitly causes an out-of-line instance to be emitted (so it doesn't have the issue we're looking at here).

`static inline` C functions aren't affected by the bug I'm fixing here as they don't have external linkage, so they don't fall into the case that does the early-out. It's still good to have this test in place though.

File: test/Interop/Cxx/function/Inputs/emit-called-function.h (hunk 7 above)
Comment (martinboehme) [reply test/Interop/Cxx/function/Inputs/emit-called-function.h]: Good point. Done.

File: lib/IRGen/GenClangDecl.cpp [hunk 8]
```diff
@@ -41,7 +41,6 @@ void IRGenModule::emitClangDecl(const clang::Decl *decl) {
   if (!valueDecl || valueDecl->isExternallyVisible()) {
     ClangCodeGen->HandleTopLevelDecl(
                           clang::DeclGroupRef(const_cast<clang::Decl*>(decl)));
-    return;
   }
```
Comment (rjmccall) [inline lib/IRGen/GenClangDecl.cpp:44]: This looks correct, but can we preserve the fast path where the declaration does not have a definition?

Comment (martinboehme) [reply lib/IRGen/GenClangDecl.cpp:46]: I assume you mean the fast path where the declaration isn't a `ValueDecl` (which I've done)?

There wasn't a fast path before for the case where the declaration doesn't have a definition (and I believe this isn't possible to check in the general case of a `ValueDecl`?).

File: test/Interop/C/function/Inputs/emit-called-function.h [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+INLINE int notCalled() {
+  return 42;
+}
+
+INLINE int calledTransitively() {
+  return 42;
+}
+
+INLINE int calledFromSwift() {
+  return calledTransitively();
```
Comment (gribozavr) [inline test/Interop/C/function/Inputs/emit-called-function.h:22]: Since there is non-trivial logic in `ClangDeclRefFinder` that distinguishes toplevel inline functions and other functions, WDYT about adding another test where we call an inline member function from a toplevel inline function?

File: test/Interop/C/function/Inputs/module.modulemap [hunk 10]
```diff
@@ -0,0 +1,4 @@
+module EmitCalledFunction {
+  header "emit-called-function.h"
```
[the remark this answers is not in the document]
Comment (martinboehme) [reply test/Interop/C/function/Inputs/module.modulemap]: Done.

File: test/Interop/C/function/Inputs/emit-called-function.h (hunk 9 above)
Comment (martinboehme) [reply test/Interop/C/function/Inputs/emit-called-function.h]: Done.

File: lib/IRGen/GenClangDecl.cpp (hunk 8 above)
Comment (rjmccall) [reply lib/IRGen/GenClangDecl.cpp:46]: Well, the fast-path was buggy, we've established that. :)  I think the point of the `isExternallyVisible()` check was supposed to be to look for a function declaration like `void foo();` that declares something with external linkage, but of course that's not the right way to check for that.

Currently the rest of the function only does special things for `FunctionDecl`s, although really it needs to do them for `VarDecl`s as well.  I think having a helper function that tries to find a body `Stmt*` for a `ValueDecl` wouldn't be remiss.

Comment (martinboehme) [reply lib/IRGen/GenClangDecl.cpp]: > Well, the fast-path was buggy, we've established that. :) I think the point of the `isExternallyVisible()` check was supposed to be to look for a function declaration like `void foo();` that declares something with external linkage, but of course that's not the right way to check for that.

Ah -- that makes sense.

> Currently the rest of the function only does special things for `FunctionDecl`s, although really it needs to do them for `VarDecl`s as well.

Oooooh, right! I've verified this case was not being handled correctly either. See the extra test case I've added -- `calledTransitivelyFromVarInit()` was not being emitted before I made the code changes in my latest commit.

> I think having a helper function that tries to find a body `Stmt*` for a `ValueDecl` wouldn't be remiss.

A `VarDecl` doesn't have a `Stmt *` though...? But I see what you mean, and I've introduced a helper function that returns a redeclaration of the decl containing "executable code" (body for a function, initializer for a variable), if one exists. This decl contains the body (in the case of a function) or initializer (in the case of a variable), so it's the right thing to traverse to find DeclRefs of things we should emit.

Not super-pleased with the naming of the helper function BTW (`getDeclWithExecutableCode()`) -- let me know if you have a better suggestion.

**Review by gribozavr — APPROVED** [review]

**Comment by gribozavr** [comment]
@swift-ci Please test

**Comment by gribozavr** [comment]
@swift-ci Please test

**Comment by gribozavr** [comment]
@swift-ci Please test macOS

**Comment by gribozavr** [comment]
@swift-ci Please test Windows

**Comment by gribozavr** [comment]
@swift-ci Please test

**Comment by gribozavr** [comment]
@swift-ci Please test

**Comment by gribozavr** [comment]
@swift-ci Please test

**Comment by gribozavr** [comment]
@swift-ci Please test

**Comment by gribozavr** [comment]
@swift-ci Please test

**Comment by gribozavr** [comment]
@swift-ci Please test

**Comment by gribozavr** [comment]
@swift-ci Please test

**Comment by gribozavr** [comment]
@swift-ci Please test macOS

**Comment by gribozavr** [comment]
@swift-ci Please test

**Comment by gribozavr** [comment]
@swift-ci Please test

**Comment by gribozavr** [comment]
@swift-ci Please test Windows

**Comment by gribozavr** [comment]
Didn't notice that we already have a recent passing Windows build: https://ci-external.swift.org/job/swift-PR-windows/2043/

Merging.

**Review by gribozavr — APPROVED** [review]

File: lib/IRGen/GenClangDecl.cpp (hunk 8 above)
Comment (rjmccall) [reply lib/IRGen/GenClangDecl.cpp]: A quirk of Clang's AST is that `Expr` is a subclass of `Stmt`, so `VarDecl`s *do* actually have a `Stmt*`, but this is a fine approach, too.

**Review by rjmccall — COMMENTED** [review]
LGTM

File: lib/IRGen/GenClangDecl.cpp (hunk 8 above)
Comment (martinboehme) [reply lib/IRGen/GenClangDecl.cpp]: Oh right, I had forgotten about that -- makes sense, given that an expression is a valid statement in C++.
